ferry/includes/importer.py

```diff
@@ -0,0 +1,770 @@
+
+        inconsistent_codes = distances_by_code[distances_by_code >= 32]
+        inconsistent_codes = list(inconsistent_codes.index)
+        inconsistent_codes = ", ".join(inconsistent_codes)
+
+        raise database.InvariantError(
+            f"Error: question codes {inconsistent_codes} have divergent texts"
+        )
+
+    print("Checking question type (narrative/rating) consistency")
+    is_narrative_by_code = evaluation_questions.groupby("question_code")[
+        "is_narrative"
+    ].apply(set)
+
+    if not all(is_narrative_by_code.apply(len) == 1):
+        inconsistent_codes = is_narrative_by_code[is_narrative_by_code.apply(len) != 1]
+        inconsistent_codes = list(inconsistent_codes.index)
+        inconsistent_codes = ", ".join(inconsistent_codes)
+        raise database.InvariantError(
+            f"Error: question codes {inconsistent_codes} have both narratives and ratings"
+        )
+
+    # deduplicate questions and keep most recent
+    evaluation_questions = evaluation_questions.sort_values(
+        by="season", ascending=False
+    )
+    evaluation_questions.drop_duplicates(
+        subset=["question_code"], keep="first", inplace=True
+    )
+
+    evaluation_questions["options"] = evaluation_questions["options"].apply(ujson.dumps)
+    evaluation_questions["options"] = evaluation_questions["options"].replace(
+        "NaN", "[]"
+    )
+
+    # explicitly specify missing columns to be filled in later
+    evaluation_statistics[["avg_rating", "avg_workload"]] = np.nan
+    evaluation_questions["tag"] = ""
+
+    # extract columns to match database
+    evaluation_narratives = evaluation_narratives.loc[
+        :, get_table_columns(database.models.EvaluationNarrative)
+    ]
+    evaluation_ratings = evaluation_ratings.loc[
+        :, get_table_columns(database.models.EvaluationRating)
+    ]
+    evaluation_statistics = evaluation_statistics.loc[
+        :, get_table_columns(database.models.EvaluationStatistics)
+    ]
+    evaluation_questions = evaluation_questions.loc[
+        :, get_table_columns(database.models.EvaluationQuestion)
+    ]
+
+    print("[Summary]")
+    print(f"Total evaluation narratives: {len(evaluation_narratives)}")
+    print(f"Total evaluation ratings: {len(evaluation_ratings)}")
+    print(f"Total evaluation statistics: {len(evaluation_statistics)}")
+    print(f"Total evaluation questions: {len(evaluation_questions)}")
+
+    return (
+        evaluation_narratives,
+        evaluation_ratings,
+        evaluation_statistics,
+        evaluation_questions,
+    )
+
+
+def copy_from_stringio(conn, df, table: str):
+    """
+    Save DataFrame in-memory and migrate
+    to database with copy_from().
+
+    Parameters
+    ----------
+    conn: psycopg2 connection object
+
+    df: DataFrame to import
+
+    table: name of target table
+
+    Returns
+    -------
+    """
+
+    # create in-memory buffer for DataFrame
+    buffer = StringIO()
+
+    df.to_csv(
```

Comment:
Can we also dump these to disk for debugging purpose?

Reply:
Sure!